Public/Src/Cache/ContentStore/Distributed/NuCache/LocalLocationStore.cs

```diff
@@ -1380,6 +1410,47 @@ public Task<BoolResult> RegisterLocalLocationAsync(OperationContext context, IRe
             }
         }
 
+        private Task<BoolResult> ProactiveReplicationAsync(OperationContext context)
+        {
+            return context.PerformOperationAsync(
+                Tracer,
+                async () =>
+                {
+                    IEnumerable<ContentInfo> localContent = await _localContentStore.GetContentInfoAsync(context.Token);
+                    localContent = localContent.OrderByDescending(info => info.LastAccessTimeUtc);
+
+                    var proactiveReplicationBlock = new ActionBlock<ContentHash>(
+                        async hash => await _proactiveReplicationTaskFactory(context, hash),
+                        new ExecutionDataflowBlockOptions()
+                        {
+                            MaxDegreeOfParallelism = _configuration.ProactiveReplicationConcurrencyLimit
+                        });
+
+                    var contentCopied = 0;
+                    foreach (var content in localContent)

```

Comment:
Maybe we need a delay between enumerations.

We can easily overwhelm the system with all the proactive replication requests happening in the entire stamp.

Reply:
Wouldn't playing with max degree of parallelism achieve the same thing?

---
In reply to: [346986895](https://github.com/microsoft/BuildXL/pull/1220#discussion_r346986895) [](ancestors = 346986895)

Reply:
Not unless there is a delay inside the proactive replication task. Suppose we want each machine to generate only 1 proactive replication per minute. You can't achieve that by setting MaxDegreeOfParallelism = 1.

---
In reply to: [347010635](https://github.com/microsoft/BuildXL/pull/1220#discussion_r347010635) [](ancestors = 347010635,346986895)